### PR TITLE
[CR] Post-threshold Mycus characters now perceive themselves as "we" in all messages, and adds framework for dynamic message changing based on conditions

### PR DESCRIPTION
**Pull request by ghost**
Hiya! I'd like to get out of the way first that I have decent experience with coding, but that this is my first time really working with C++. I included the CR tag because I don't know if this is a good way to do this, or if it would be significantly slower than what exists now. If this is a much higher amount of CPU/memory use just for one fluff thing and it can't be easily smoothed out, I can close it.

The alternative that occurs to me is replacing all the messages in the game with descriptors, i.e. `You eat your %s.` becomes `$YOU eat $your %s.`, with the $YOU and $your replaced dynamically based on conditions. The obvious problem there is that, well, I'd have to replace every single message in the game. It's a massive amount of work that I'm not sure would even be accepted for something so small.

Anyways, this does what it says on the tin. Because the Mycus has no sense of individuality, it always refers to itself as "we", "our", etc., and in actions specific to the Mycus mutations, it shows (i.e. "We can't eat that. It's not for us.") This PR makes it so that all messages the player receives are dynamically changed to reflect that if they've become a local guide for the Mycus.

In the former half of this image, the player has no mutations; in the latter half, they are Mycus:
before and after

This also applies to things such as combat.
hot mushroom on zombie action

To accomplish this, whenever a message is to be outputted to the player, it's run through a new function called `parse_replacements` that replaces words or phrases in the message based on conditions via another new function, `replace_text`. (There might already be a similar function for either of these, but I'm new to CDDA's code so if there is I'm unaware of it. Please let me know!) Right now, `parse_replacements` only checks for post-threshold Mycus, but adding new conditions and replacements is easy.

Because the message swapping is dynamic, there are some edge cases that'll stand out. For instance, because the lowercase form of "you" is used in separate ways - both "if only you had a crowbar" and "the zombie grabs you" stand out - the edge cases I'm aware of are just handled directly.

Thoughts, feedback, ideas, etc. all welcome. Again, I've no real experience with C++, so my code probably leaves something to be desired.

### CODE REVIEW

**Comment by StripedMonkey** [comment]
I personally think that this is a pretty interesting change, in terms of implementation however I don't think this works with localization for other languages, although I'm not sure if making it easily localized is even possible.

**Comment by Xhuis** [comment]
I'm not actually sure how localization is handled. It sounds fairly simple to just check and replace translated words in the other locals, but I'm not sure of how to implement it.

**Comment by StripedMonkey** [comment]
Making the change to have all messages have $pronoun would not be difficult and I think it would give the ability for others to make similar changes more easily. It would be a matter of find/replace but I'm not sure if that would be approved. It would certainly have less overhead than find/replace each message but as this is just a message output I don't think its that big of a worry.

**Comment by cainiaowu** [comment]
There are "snippet-like" system which use tag to create sentences (like swearing for one) in the code somewhere, which might help this.
You use one set for the tag if the Player has Mycus, and you use the other set if not.

**Comment by Xhuis** [comment]
Ahh, so the message replacements would be stored in the json, and read/applied from there instead? I don't entirely follow.

**Comment by kevingranade** [comment]
Cool idea, what we've done to handle this kind of thing in the past is to replace the word with a tag, like, "<npcname>" in literal strings, and run it through a specialized method like add_msg_player_or_npc() to apply the substitution.
Just as importantly, the tags and the strings that are substituted for the tags are captured by the translation system.

The tag based approach might not work for some languages due to ordering and conjugation, so there's a different approach we could use, which is to apply your replacements using macros applied at compile time to produce both variants of the strings in question, then translators would be free to translate and conjugate the full phrases.  I'd like to hear whether that's necessary though, because it would also mean more work for translators.

https://github.com/CleverRaven/Cataclysm-DDA/blob/4c0b3eee36d5b9f9c42ac4a898bdf17effb855e8/src/item_factory.cpp#L723

**Comment by cainiaowu** [comment]
The current tag system is useful enough, at least for Chinese translation.

**Comment by Xhuis** [comment]
The tag system is the alternative that occurred to me, but I wasn't sure if it was preferable to the macro system that I implemented. It was also the quicker route, as otherwise I would have to replace every message that refers to the player with a tag, and such formatting would have to be upheld for future messages so that they'd properly change.

Should using tags be by and large the preferred option, I could migrate to it, although it could cause plenty of merge conflicts with any other PRs that modify messages.

**Comment by Xhuis** [comment]
Just to make sure I'm not doing anything incorrectly, should I be replacing all messages with the tag system instead of using the macro system like I am now? I'm not certain if it was an instruction or just discussing the merits between the two.

**Comment by kevingranade** [comment]
If it allows for proper translation, use the tags system in add_msg_player_or_npc().
We have at least one point of feedback that its sufficient, in Chinese even, so I'd say to go with that approach.

It's also a bad code smell that the messages module depends on the player module.

**Comment by Xhuis** [comment]
Okie doke. I agree that it's probably not that good of code; I'm still getting a handle for how DDA ties itself together to make it what it is. I'll get it migrated when I can.

**Comment by Xhuis** [comment]
I've run into a roadblock, so I figured I'd ask here.

It seems that the function `replace_with_npc_name` (in this case, I plan to rename it to `replace_tags`) is innate to the Creature namespace, and is a protected function. Trying to simply reuse it in the `player.cpp` functions for handling messages doesn't work - for instance, I changed the pickup message to:

`<you> pick up (...)`
instead of
`You pick up (...)`

and then, in the `replace_tags` definition in `creature.cpp`, added the following:

`replace_substring( input, "<you>", "You", true );`

Which resulted in the message simply saying ` pick up the pocket knife`, for instance, and erasing the tag entirely. I'm not sure how best to work around this, and my knowledge of C++ is too poor for me to accurately diagnose the issue after some hours of work; should I be migrating it to a different namespace other than `Creature`? Maybe switching it from `protected` to `public` and redefining it for the `player` namespace? All suggestions are welcome; I'm still reading on C++, and I would likely figure it out given time, but help is always appreciated to get it done sooner.

File: src/messages.cpp [hunk 1]
```diff
@@ -214,6 +216,32 @@ void Messages::deserialize( JsonObject &json )
     obj.read( "curmes", player_messages.impl_->curmes );
 }
 
+std::string Messages::replace_text( std::string& str, const std::string& from, const std::string& to )
+{
+    size_t start_pos = str.find(from);
+    if(start_pos == std::string::npos)
```
Comment (BevapDin) [inline src/messages.cpp]: Please follow the code style guide (see doc/CODE_STYLE.md). In this case: spaces after `(` and before `)` and brackets around the following block.

**Comment by Xhuis** [comment]
I do still like the concept of this PR, but I haven't had a chance to dive into C++ more to observe its workings nor to work on the issues present here. I don't want to clog the tracker, so in a few days once I get home, if I still can't find a solution, I'll close this.

**Comment by Xhuis** [comment]
Luck has still evaded me in discovering a solution to this, so it'll have to wait until I know more C++. Thanks for the support, though, y'all! Once I have a better handle on the language it should come easy to me.

**Pull request closed without merging**